src/App.tsx

```diff
@@ -1,61 +1,123 @@
+  const [hasTitleError, setHasTitleError] = useState(false);
+  const [hasUserIdError, setHasUserIdError] = useState(false);
+
+  const getTodoWithUsers = () => {
+    return todos.map(todo => {
+      const user = usersFromServer.find(
+        currUser => currUser.id === todo.userId,
+      );
+
+      if (!user) {
+        throw new Error('!user');
+      }
+
+      return {
+        ...todo,
+        user,
+      };
+    });
+  };
+
+  const onTitleChange = (event: React.ChangeEvent<HTMLInputElement>) => {
+    setTitle(event.target.value);
+    setHasTitleError(false);
+  };
+
+  const onUserIdChange = (event: React.ChangeEvent<HTMLSelectElement>) => {
+    setUserId(+event.target.value);
+    setHasUserIdError(false);
+  };
+
+  const onSubmit = (newTodo: Todo) => {
+    const newId = Math.max(...todos.map(todo => todo.id)) + 1;
+    const currentTodo = {
+      ...newTodo,
+      id: newId,
+    };
+
+    setTodos(currentTodos => [...currentTodos, currentTodo]);
+  };
+
+  const handleSubmit = (event: React.FormEvent) => {
+    event.preventDefault();
+
+    setHasTitleError(!title);
+    setHasUserIdError(!userId);
+
+    if (!title || !userId) {
+      return;
+    }
+
+    onSubmit({
+      id: 0,
+      title: title,
+      completed: false,
+      userId: userId,
+      user: usersFromServer.find(user => user.id === userId) || null,
+    });
+
+    setTitle('');
+    setUserId(0);
+  };
+
   return (
     <div className="App">
       <h1>Add todo form</h1>
 
       <form action="/api/todos" method="POST">
         <div className="field">
-          <input type="text" data-cy="titleInput" />
-          <span className="error">Please enter a title</span>
+          <label htmlFor="title">Title:&nbsp;</label>
+          <input
+            id="title"
+            type="text"
+            data-cy="titleInput"
+            placeholder="Enter a title"
+            value={title}
+            onChange={onTitleChange}
+          />
+          {hasTitleError && <span className="error">Please enter a title</span>}
         </div>
 
         <div className="field">
-          <select data-cy="userSelect">
+          <label htmlFor="user">User:&nbsp;</label>
+          <select
+            id="user"
+            data-cy="userSelect"
+            value={userId}
+            onChange={onUserIdChange}
+          >
             <option value="0" disabled>
               Choose a user
             </option>
+            {usersFromServer.map(currUser => (
+              <option key={currUser.id} value={currUser.id}>
+                {currUser.name}
+              </option>
+            ))}
           </select>
 
-          <span className="error">Please choose a user</span>
+          {hasUserIdError && (
+            <span className="error">Please choose a user</span>
+          )}
         </div>
 
-        <button type="submit" data-cy="submitButton">
+        <button type="submit" data-cy="submitButton" onClick={handleSubmit}>
```

Comment:
Remove the `onClick` event handler from the submit button and move the `handleSubmit` function to the form's `onSubmit` event.